### PR TITLE
Undo redo

### LINKED ISSUE #412: Undo
In talking with people in the lab, the major use case for undo is to undo either an accidental addition or deletion of an annotation. I am assuming undo addition is easier to solve than undo deletion. It would also be helpful to undo annotations that are introduced in a chunk from, say, an annotation worker. I don't think anybody was looking for a full version history-level undo feature. Just the last step would be fine.

### PR DESCRIPTION
**Pull request by bruyeret**
Add Undo/Redo feature
Add an history to the database which contains history entries used to undo/redo recordable actions.

Mechanism to record an API call:
- All models in the database inherit `ProxiedAccessControlledModel` (except for History)
- The endpoint method is decorated with `@recordable(...)` (and eventually `@cacheBodyJson`)
- The recordable decorator first call a girder event starting the recording for all `ProxiedAccessControlledModel`
- It calls the original endpoint function ; the calls of the endpoint function to models method (`save`, `remove`, ...) are recorded by the `ProxiedAccessControlledModel`
- It calls a girder event stoping the recording for all `ProxiedAccessControlledModel` and gathering the records
- It adds the history entry to the database

Close #412

### CODE REVIEW

File: devops/girder/plugins/AnnotationPlugin/upenncontrast_annotation/server/api/annotation.py [hunk 1]
```diff
@@ -1,9 +1,42 @@
...
+    annotationStringIds = self.getBodyJson()
+    query = {
+      '_id': { '$in': [ObjectId(stringId) for stringId in annotationStringIds] },
+    }
+    cursor = self._annotationModel.findWithPermissions(query, user=self.getCurrentUser(), level=AccessType.READ, limit=1)
+    annotation = next(cursor, None)
+    if annotation is None:
+        return None
+    else:
+        return annotation['datasetId']
```
Comment (luciemac) [inline devops/girder/plugins/AnnotationPlugin/upenncontrast_annotation/server/api/annotation.py:39]: `return annotation is None ? None: annotation['datasetId']` ?

File: devops/girder/plugins/AnnotationPlugin/upenncontrast_annotation/server/api/annotation.py [hunk 2]
```diff
@@ -1,9 +1,39 @@
...
+
+def getDatasetIdFromAnnotationInBody(self: 'Annotation', *args, **kwargs):
+    annotation = self.getBodyJson()
+    return annotation['datasetId']
+
+def getDatasetIdFromAnnotationListInBody(self: 'Annotation', *args, **kwargs):
+    annotations = self.getBodyJson()
+    if len(annotations) <= 0:
+        return None
+    else:
```
Comment (luciemac) [inline devops/girder/plugins/AnnotationPlugin/upenncontrast_annotation/server/api/annotation.py:22]: this else is not necessary

File: devops/girder/plugins/AnnotationPlugin/upenncontrast_annotation/server/api/connections.py [hunk 3]
```diff
@@ -1,9 +1,51 @@
...
+
+def getDatasetIdFromConnectionInBody(self: 'AnnotationConnection', *args, **kwargs):
+    connection = self.getBodyJson()
+    return connection['datasetId']
+
+def getDatasetIdFromConnectionListInBody(self: 'AnnotationConnection', *args, **kwargs):
+    connections = self.getBodyJson()
+    if len(connections) <= 0:
+        return None
+    else:
```
Comment (luciemac) [inline devops/girder/plugins/AnnotationPlugin/upenncontrast_annotation/server/api/connections.py:23]: same here

**Comment by luciemac** [comment]
LGTM

**Comment by arjunrajlab** [comment]
Just tested, works great! Minor request: can you tie undo to Cmd-Z (Apple) and Control-Z (Windows/Linux) and redo to Cmd-shift-Z (Apple) and Control-shift-Z (Windows/Linux)?

**Comment by bruyeret** [comment]
It should be ready to merge now, I added the shortcut

**Comment by luciemac** [comment]
LGTM

**Pull request merged**